pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -0,0 +1,336 @@
+	newObject, update, err := strategy.ShouldUpdate(current, bootstrap)
+	if err != nil {
+		return fmt.Errorf("failed to determine whether auto-update is required for %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	}
+	if !update {
+		if klog.V(5).Enabled() {
+			// TODO: if we use structured logging here the diff gets escaped and very awkward to read in the log
+			klog.Infof("No update required for the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, bootstrap))
+		}
+		return nil
+	}
+
+	if _, err := wrapper.Update(newObject); err != nil {
+		return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	}
+
+	klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, newObject))
+	return nil
+}
+
+func removeConfiguration(wrapper configurationWrapper, bootstrap metav1.Object) error {
+	name := bootstrap.GetName()
+
+	current, err := wrapper.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to retrieve the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
+	}
+
+	value := current.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey]
+	autoUpdate, err := strconv.ParseBool(value)
+	if err != nil {
+		klog.ErrorS(err, fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
+
+		// This may need manual intervention, in case the annotation value is malformed,
+		// so don't return an error, that might trigger futile retry loop.
+		return nil
+	}
+	if !autoUpdate {
+		klog.V(5).InfoS(fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
+		return nil
+	}
+
+	if err := wrapper.Delete(name); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to delete the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
+	}
+
+	klog.V(2).InfoS(fmt.Sprintf("Successfully deleted the %s", wrapper.TypeName()), "name", name)
+	return nil
+}
+
+// getRemoveCandidate returns a list of configuration objects we should delete
+// from the cluster given a set of bootstrap and current configuration.
+// bootstrap: a set of hard coded configuration kube-apiserver maintains in-memory.
+// current: a set of configuration objects that exist on the cluster
+// Any object present in current is a candidate for removal if both a and b are true:
+//  a. the object in current is missing from the bootstrap configuration
+//  b. the object has the designated auto-update annotation key
+// This function shares the common logic for both FlowSchema and PriorityLevelConfiguration
+// type and hence it accepts metav1.Object only.
+func getRemoveCandidate(bootstrap map[string]metav1.Object, current []metav1.Object) []metav1.Object {
+	if len(current) == 0 || len(bootstrap) == 0 {
```

Comment:
This isn't correct, right?

If bootstrap is empty, then we may potentially want to remove stuff (that's not there).

[would be useful to add a test for it]

Reply:
yes, theoretically bootstrap can be empty.